core/src/main/java/org/infinispan/persistence/util/PersistenceManagerCloseableSupplier.java

```diff
@@ -0,0 +1,159 @@
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+/**
+ * A closeable supplier that provides a wait to supply cache entries from a given persistence manager.  On the first
+ * call to get this class will submit a task to collect all of the entries from the loader (or optionally a subset
+ * provided a given {@link org.infinispan.filter.KeyFilter}).  A timeout value is required so that if a get blocks
+ * for the given timeout it will throw a {@link TimeoutException}.
+ * @author William Burns
+ * @since 8.0
+ */
+public class PersistenceManagerCloseableSupplier<K, V> implements CloseableSupplier<CacheEntry<K, V>> {
+   private final Executor executor;
+   private final PersistenceManager manager;
+   private final KeyFilter<K> filter;
+   private final InternalEntryFactory factory;
+   private final BlockingQueue<CacheEntry<K, V>> queue;
+   private final long timeout;
+   private final TimeUnit unit;
+
+   private final Lock closeLock = new ReentrantLock();
+   private final Condition closeCondition = closeLock.newCondition();
+
+   private boolean closed = false;
+   private AtomicReference<AdvancedCacheLoader.CacheLoaderTask<K, V>> taskRef = new AtomicReference<>();
+
+   public PersistenceManagerCloseableSupplier(Executor executor, PersistenceManager manager,
+                                              InternalEntryFactory factory, KeyFilter<K> filter, long timeout,
+                                              TimeUnit unit, int maxQueue) {
+      this.executor = executor;
+      this.manager = manager;
+      this.factory = factory;
+      this.filter = filter;
+      this.timeout = timeout;
+      this.unit = unit;
+      this.queue = new ArrayBlockingQueue<>(maxQueue);
+   }
+
+   class SupplierCacheLoaderTask implements AdvancedCacheLoader.CacheLoaderTask<K, V> {
+
+      @Override
+      public void processEntry(MarshalledEntry<K, V> marshalledEntry, AdvancedCacheLoader.TaskContext taskContext)
+              throws InterruptedException {
+         if (!taskContext.isStopped()) {
+            closeLock.lock();
+            try {
+               if (closed) {
+                  taskContext.stop();
+                  return;
+               }
+            } finally {
+               closeLock.unlock();
+            }
+            InternalCacheEntry<K, V> ice = PersistenceUtil.convert(marshalledEntry, factory);
+            // We do a read without acquiring lock
+            boolean stop = closed;
+            while (!stop) {
+               // If we were able to offer a value this means someone took from the queue so let us come back around main
+               // loop to offer all values we can
+               // TODO: do some sort of batching here - to reduce wakeup contention ?
+               if (queue.offer(ice, 100, TimeUnit.MILLISECONDS)) {
```

Comment:
Yeah tbh this is a part of code I want to rework at some point to make a bit more efficient.  Right now it can get in an ugly wakeup lock contention as a get will block waiting for a new value and the task adds it going back and forth if the value is consumed from supplier faster than they are populated.

Unfortunately ArrayDeque is not bounded and I would have to find a way to wait on the consumer in that case which is why I use a bounded blocking queue.  But I am open to any suggestions.

My thought of refactoring this is that the task would hold a collection internally say an ArrayList that is sized to maxQueue / 4.  Then instead of pushing directly to the queue it would push to the array list instead.  And only when the array list has gotten to maxQueue / 4 would it check the task being stopped and finally push it into the blocking queue.  This would limit lock acquisition significantly and also the consumer wouldn't get stuck having to wake up acquire lock constantly.